server/src/main/java/io/crate/replication/logical/action/ShardChangesAction.java

```diff
@@ -231,6 +237,33 @@ protected ShardsIterator shards(ClusterState state, Request request) {
                 request.shardId().id()
             ).activeInitializingShardsRandomIt();
         }
+
+        /**
+         * A publisher should not send updates when a subscriber polls changes of an index
+         * where not all shards are active yet, restore will fail if primaries are not (yet) assigned.
+         * <p>
+         * We include such indices to the PublicationsStateAction.Response as otherwise subscriber
+         * can treat "not ready to replicate" index as deleted and turn a table into a regular one on its side.
+         * For example, this can happen on a publisher cluster restart.
+         *
+         * @return boolean flag indicating that listener is failed and async callee can return right away.
+         */
+        private boolean ensureAllPrimariesActive(ClusterState state,
+                                                 ShardId shardId,
+                                                 @Nullable ActionListener<Response> listener) {
+            var routingTable = state.routingTable().index(shardId.getIndexName());
+            if (routingTable != null && routingTable.allPrimaryShardsActive() == false) {
```

Comment:
`routingTable` can be null on publisher (on debugging there is no table corresponding to `shardId.getIndexName(`)) not sure if it's related to test setup or normal situation. 

On master when polling is done against such state, we fail on   
`var indexShard = indicesService.indexServiceSafe(shardId.getIndex()).getShard(shardId.id()); `with `IndexNotFoundException` - we don't fail listener but it's get caught somewhere in `InboundHandler` and then listener fails there. 

this btw means we can just throw on async execution as well - but I decided to shortcut a bit instead of relying on further handling in the `InboundHandler`